File: src/StateVector.ts
import { OpId } from "./OpId";
import type { OpIdRange } from "./treeTypes";
import type { NodeOperation } from "./operations";

/**
 * Helper function to subtract one set of ranges from another.
 * Returns the ranges in A that are not in B.
 * Assumes ranges in both A and B are sorted and non-overlapping.
 */
export function subtractRanges(rangesA: number[][], rangesB: number[][]): number[][] {
  if (rangesB.length === 0) return rangesA.map(r => [...r]); // Return a copy
  if (rangesA.length === 0) return []; // If A is empty, nothing to subtract

  const result: number[][] = [];
  let indexB = 0;

  for (const rangeA of rangesA) {
    let currentStart = rangeA[0];
    const endA = rangeA[1];

    // Iterate through ranges in B that could potentially overlap with rangeA
    while (indexB < rangesB.length && rangesB[indexB][1] < currentStart) {
      // Skip ranges in B that are entirely before the current start
      indexB++;
    }

    while (indexB < rangesB.length && rangesB[indexB][0] <= endA) {
      const startB = rangesB[indexB][0];
      const endB = rangesB[indexB][1];

      // If there's a gap before this range in B starts
      if (currentStart < startB) {
        // Add the portion of rangeA before the overlap
        result.push([currentStart, Math.min(endA, startB - 1)]);
      }

      // Move current start past this range in B
      currentStart = Math.max(currentStart, endB + 1);

      // If we've gone past the end of rangeA, break inner loop
      if (currentStart > endA) break;

      // If the current rangeB ends after rangeA, we don't need to check further ranges in B for this rangeA
      if (endB >= endA) break;

      // Only advance indexB if the current rangeB is fully processed relative to currentStart
      if (endB < currentStart) {
        indexB++;
      } else if (startB >= currentStart) {
        indexB++;
      }
    }

    // If there's a remaining part of rangeA after processing overlaps with B
    if (currentStart <= endA) {
      result.push([currentStart, endA]);
    }
  }

  return result;
}

/**
 * StateVector tracks operations that have been applied using a range-based representation.
 * It's used for synchronization between peers to determine which operations need to be sent.
 */
export class StateVector {
  private ranges: Record<string, number[][]> = {};

  /**
   * Creates a new StateVector.
   * @param initialState Optional initial state to copy from
   */
  constructor(initialState: Record<string, number[][]> = {}) {
    // Create a deep copy of the initial state
    for (const [peerId, peerRanges] of Object.entries(initialState)) {
      this.ranges[peerId] = peerRanges.map(range => [...range]);
    }
  }

  /**
   * Updates the state vector with a newly applied operation.
   * Assumes ranges are sorted and non-overlapping.
   *
   * @param peerId The peer ID of the operation
   * @param counter The counter value of the operation
   */
  update(peerId: string, counter: number): void {
    // Initialize ranges array for this peer if it doesn't exist
    if (!this.ranges[peerId]) {
      this.ranges[peerId] = [];
    }

    const ranges = this.ranges[peerId];

    // Case 1: No ranges yet
    if (ranges.length === 0) {
      ranges.push([counter, counter]);
      return;
    }

    let rangeExtendedOrMerged = false;
    let insertIndex = -1;

    for (let i = 0; i < ranges.length; i++) {
      const range = ranges[i];

      // If counter is already in a range, do nothing
      if (counter >= range[0] && counter <= range[1]) {
        rangeExtendedOrMerged = true;
        break;
      }

      // If counter is one less than range start, extend range start
      if (counter === range[0] - 1) {
        range[0] = counter;
        rangeExtendedOrMerged = true;
        // Check if this range now merges with the previous range
        if (i > 0 && range[0] === ranges[i - 1][1] + 1) {
          ranges[i - 1][1] = range[1]; // Merge into previous
          ranges.splice(i, 1); // Remove current
        }
        break;
      }

      // If counter is one more than range end, extend range end
      if (counter === range[1] + 1) {
        range[1] = counter;
        rangeExtendedOrMerged = true;
        // Check if this range now merges with the next range
        if (i < ranges.length - 1 && range[1] + 1 === ranges[i + 1][0]) {
          range[1] = ranges[i + 1][1]; // Merge next into current
          ranges.splice(i + 1, 1); // Remove next
        }
        break;
      }

      // Keep track of where to insert if no extension/merge happens
      if (counter < range[0] && insertIndex === -1) {
        insertIndex = i;
      }
    }

    // If we couldn't extend or merge any range, add a new one
    if (!rangeExtendedOrMerged) {
      if (insertIndex === -1) {
        // If counter is greater than all existing ranges, add to the end
        insertIndex = ranges.length;
      }
      ranges.splice(insertIndex, 0, [counter, counter]);
      // After inserting, check if the new range merges with neighbors
      // Merge with previous range if possible
      if (insertIndex > 0 && ranges[insertIndex][0] === ranges[insertIndex - 1][1] + 1) {
        ranges[insertIndex - 1][1] = ranges[insertIndex][1];
        ranges.splice(insertIndex, 1);
        insertIndex--; // Adjust index after merging
      }
      // Merge with next range if possible (use adjusted insertIndex)
      if (insertIndex < ranges.length - 1 && ranges[insertIndex][1] + 1 === ranges[insertIndex + 1][0]) {
        ranges[insertIndex][1] = ranges[insertIndex + 1][1];
        ranges.splice(insertIndex + 1, 1);
      }
    }
  }

  /**
   * Updates the state vector with a newly applied operation.
   *
   * @param op The operation that was just applied
   */
  updateFromOp(op: NodeOperation): void {
    this.update(op.id.peerId, op.id.counter);
  }

  /**
   * Removes a single operation ID from the state vector.
   * Assumes ranges are sorted and non-overlapping.
   *
   * @param peerId The peer ID of the operation
   * @param counter The counter value of the operation
   * @returns true if a counter was removed, false if it was absent
   */
  remove(peerId: string, counter: number): boolean {
    const ranges = this.ranges[peerId];
    if (!ranges) {
      return false;
    }

    for (let i = 0; i < ranges.length; i++) {
      const range = ranges[i];
      const [start, end] = range;

      if (counter < start) {
        return false;
      }

      if (counter > end) {
        continue;
      }

      if (start === end) {
        ranges.splice(i, 1);
      } else if (counter === start) {
        range[0] = start + 1;
      } else if (counter === end) {
        range[1] = end - 1;
      } else {
        ranges.splice(i, 1, [start, counter - 1], [counter + 1, end]);
      }

      if (ranges.length === 0) {
        delete this.ranges[peerId];
      }
      return true;
    }

    return false;
  }

  /**
   * Removes the operation ID from the state vector.
   *
   * @param op The operation to remove
   * @returns true if the operation ID was removed, false if it was absent
   */
  removeFromOp(op: NodeOperation): boolean {
    return this.remove(op.id.peerId, op.id.counter);
  }

  /**
   * Returns the current state vector.
   * Returns a deep copy so callers cannot mutate internal state.
   */
  getState(): Readonly<Record<string, number[][]>> {
    const state: Record<string, number[][]> = {};
    for (const [peerId, peerRanges] of Object.entries(this.ranges)) {
      state[peerId] = peerRanges.map(range => [...range]);
    }
    return state;
  }

  /**
   * Calculates which operation ranges we have that the other state vector is missing
   * by comparing state vectors.
   *
   * @param other The other state vector to compare against
   * @returns Array of operation ID ranges that we have but they don't
   */
  diff(other: StateVector): OpIdRange[] {
    const missingRanges: OpIdRange[] = [];

    // Check what we have that they don't have
    for (const [peerId, ourRanges] of Object.entries(this.ranges)) {
      const theirRanges = other.ranges[peerId] || [];

      // Calculate ranges we have that they don't
      const missing = subtractRanges(ourRanges, theirRanges);

      // Convert to OpIdRange format
      for (const [start, end] of missing) {
        // Ensure the range is valid (start <= end)
        if (start <= end) {
          missingRanges.push({
            peerId,
            start,
            end
          });
        }
      }
    }

    return missingRanges;
  }

  /**
   * Checks if the state vector contains the given operation ID
   *
   * @param opId The operation ID to check
   * @returns true if the operation is in the state vector, false otherwise
   */
  contains(opId: OpId): boolean {
    const peerId = opId.peerId;
    const counter = opId.counter;

    if (!this.ranges[peerId]) {
      return false;
    }

    for (const [start, end] of this.ranges[peerId]) {
      if (counter >= start && counter <= end) {
        return true;
      }
    }

    return false;
  }

  /**
   * Creates a copy of this state vector
   */
  clone(): StateVector {
    return new StateVector(this.ranges);
  }

  /**
   * Builds a state vector from an array of operations
   * @param operations The operations to build the state vector from
   * @returns A new StateVector instance
   */
  static fromOperations(operations: ReadonlyArray<NodeOperation>): StateVector {
    const stateVector = new StateVector();
    for (const op of operations) {
      stateVector.updateFromOp(op);
    }
    return stateVector;
  }
}

File: reptree-rs/src/crdt/state_vector.rs
//! State vector implementation for RepTree
//!
//! This module provides a range-based state vector implementation for tracking
//! which operations have been applied across peers.

use crate::types::{OpId, Range};
use std::collections::HashMap;

/// State vector for tracking applied operations
pub struct StateVector {
    /// Map of peer ID to ranges of applied operations
    ranges: HashMap<String, Vec<Range>>,
}

impl StateVector {
    /// Create a new empty state vector
    pub fn new() -> Self {
        Self {
            ranges: HashMap::new(),
        }
    }
    
    /// Create a state vector from a map of ranges
    pub fn from_ranges(ranges: HashMap<String, Vec<Range>>) -> Self {
        Self { ranges }
    }
    
    /// Add an operation to the state vector
    pub fn add(&mut self, peer_id: &str, counter: u64) {
        let ranges = self.ranges.entry(peer_id.to_string()).or_insert_with(Vec::new);
        
        // Find if we can extend an existing range
        let mut extended = false;
        for range in ranges.iter_mut() {
            // If the counter is adjacent to the end of the range, extend it
            if range.end + 1 == counter {
                range.end = counter;
                extended = true;
                break;
            }
            // If the counter is adjacent to the start of the range, extend it
            else if counter + 1 == range.start {
                range.start = counter;
                extended = true;
                break;
            }
            // If the counter is already in the range, do nothing
            else if counter >= range.start && counter <= range.end {
                extended = true;
                break;
            }
        }
        
        // If we couldn't extend an existing range, create a new one
        if !extended {
            ranges.push(Range {
                peer_id: peer_id.to_string(),
                start: counter,
                end: counter,
            });
        }
        
        // Sort and merge overlapping ranges
        self.normalize_ranges(peer_id);
    }
    
    /// Sort and merge overlapping ranges for a peer
    fn normalize_ranges(&mut self, peer_id: &str) {
        if let Some(ranges) = self.ranges.get_mut(peer_id) {
            // Sort ranges by start
            ranges.sort_by_key(|r| r.start);
            
            // Merge overlapping ranges
            let mut i = 0;
            while i < ranges.len() - 1 {
                let current = &ranges[i];
                let next = &ranges[i + 1];
                
                // If the ranges overlap or are adjacent, merge them
                if current.end + 1 >= next.start {
                    let merged = Range {
                        peer_id: peer_id.to_string(),
                        start: current.start,
                        end: std::cmp::max(current.end, next.end),
                    };
                    
                    ranges[i] = merged;
                    ranges.remove(i + 1);
                } else {
                    i += 1;
                }
            }
        }
    }
    
    /// Get the ranges for all peers
    pub fn get_ranges(&self) -> HashMap<String, Vec<Range>> {
        self.ranges.clone()
    }
    
    /// Calculate the difference between this state vector and another
    pub fn diff(&self, other: &Self) -> Vec<Range> {
        let mut result = Vec::new();
        
        // For each peer in our state vector
        for (peer_id, our_ranges) in &self.ranges {
            // Get the ranges for this peer in the other state vector
            let their_ranges = other.ranges.get(peer_id).unwrap_or(&Vec::new());
            
            // Calculate the ranges we have that they don't
            for our_range in our_ranges {
                let mut remaining = vec![our_range.clone()];
                
                for their_range in their_ranges {
                    let mut new_remaining = Vec::new();
                    
                    for range in remaining {
                        // If the range is completely before their range, keep it
                        if range.end < their_range.start {
                            new_remaining.push(range);
                        }
                        // If the range is completely after their range, keep it
                        else if range.start > their_range.end {
                            new_remaining.push(range);
                        }
                        // If the range overlaps with their range, split it
                        else {
                            // If there's a part before their range, keep it
                            if range.start < their_range.start {
                                new_remaining.push(Range {
                                    peer_id: peer_id.clone(),
                                    start: range.start,
                                    end: their_range.start - 1,
                                });
                            }
                            
                            // If there's a part after their range, keep it
                            if range.end > their_range.end {
                                new_remaining.push(Range {
                                    peer_id: peer_id.clone(),
                                    start: their_range.end + 1,
                                    end: range.end,
                                });
                            }
                        }
                    }
                    
                    remaining = new_remaining;
                }
                
                // Add the remaining ranges to the result
                result.extend(remaining);
            }
        }
        
        result
    }
}
